README.md
# fp-ssc-course

### An introduction to functional programming for scalable statistical computing and machine learning

## A half-day short-course

A brief introduction to ideas of functional programming in the context
of scalable statistical computing, illustrated with hands-on examples
in Scala, Haskell, (Python+)JAX and
Dex. [Scala](https://www.scala-lang.org/) and
[Haskell](https://www.haskell.org/) are general purpose strongly typed
functional programming
languages. [JAX](https://jax.readthedocs.io/en/latest/) is a
functional language for differentiable array programming embedded in
Python, and [Dex](https://github.com/google-research/dex-lang) is a
new experimental strongly typed functional language for differentiable
array processing.

**This course is still being developed.** The first iteration of this course will be delivered to [StatML](https://statml.io/) PhD students on 2023-02-16.

*Please note that you need to install some software on your system **in advance** of the course.* See the [Setup](Setup.md) guide for details.

You will also need a copy of this repo. If you know git, clone it ASAP, and then do a pull the day before the course. If you don't know git, click on the green "Code" button and download a zip file, but do (or re-do) this the day before the course to make sure you have an up-to-date version.

It would also be useful to have a copy of my [logreg](https://github.com/darrenjw/logreg) repo, which explores MCMC algorithms in a variety of (functional) languages.

### Abstract

Non-trivial research problems in statistical computing and machine
learning are often complex and computationally intensive, requiring a
custom implementation in some programming language. All of the
languages commonly used for this purpose are very old, dating back to
the dawn of the computing age, and are quite unsuitable for scalable
and efficient statistical computation. There have been huge advances in
computing science in the decades since these languages were created,
and many new, different and better programming languages have been
created. Although functional programming languages are not new, there
has been a large resurgence of interest in functional languages in the
last decade or two, as people have begun to appreciate the advantages
of the functional approach, especially in the context of developing
large, scalable software systems, and the ability to take advantage of
modern computing hardware.

This short course will provide a brief introduction to ideas of
functional programming in the context of scalable statistical
computing, illustrated with hands-on examples in Scala, Haskell,
(Python+)JAX and Dex. Scala and Haskell are general purpose strongly
typed functional programming languages. JAX is a functional language
for differentiable array programming embedded in Python, and Dex is a
new experimental strongly typed functional language for differentiable
array processing.

### Materials

* [*Introduction to FP*](Intro/Readme.md)
* [*Scala crash-course*](Scala/md/ScalaCC.md)
* [Scala hands-on](Scala/md/ScalaHO.md)
* [*Running example*](Intro/Example.md)
* [Scala example and hands-on](Scala/md/Example.md)
* Haskell crash-course
* [Haskell example and hands-on](Haskell/Example.md)
* JAX crash-course
* [JAX example and hands-on](JAX/Example.md)
* Dex crash-course
* [Dex example and hands-on](Dex/Example.md)
* [Functional and parallel random numbers](Intro/Random.md)
* Splittable random numbers hands-on
* [Wrap-up and next steps](Intro/Resources.md)




Haskell/CHANGELOG.md
# Changelog for `ml-ga`

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to the
[Haskell Package Versioning Policy](https://pvp.haskell.org/).

## Unreleased

## 0.1.0.0 - YYYY-MM-DD


Scala/src/test/scala/ML-GA-pbt.scala
import munit.ScalaCheckSuite
import org.scalacheck.Prop.*

import cats.*
import cats.implicits.*

// Example property-based tests

class MyPropertyTests extends ScalaCheckSuite:

  property("An Int should combine commutatively") {
    forAll { (a: Int, b: Int) =>
      assertEquals(a |+| b, b |+| a)
    }
  }

  property("An Int should invert") {
    forAll { (a: Int) =>
      assertEquals(a |+| a.inverse(), Monoid[Int].empty)
    }
  }


// eof




Scala/src/test/scala/ML-GA-laws.scala
import cats.*
import cats.implicits.*

import munit.DisciplineSuite

// Example Cats laws tests

class MyLawTests extends DisciplineSuite:

  import cats.kernel.laws.discipline.SemigroupTests

  checkAll("List[Int].SemigroupLaws", SemigroupTests[List[Int]].semigroup)

  import cats.laws.discipline.FunctorTests

  checkAll("List.FunctorLaws", FunctorTests[List].functor[Int, Int, String])

  import cats.laws.discipline.MonadTests
  // checking monad laws is quite slow
  //checkAll("List.MonadLaws", MonadTests[List].monad[Int, Int, String])



// eof




Scala/src/test/scala/ML-GA-test.scala
import org.scalatest.flatspec.AnyFlatSpec

class SetSpec extends AnyFlatSpec:

  import breeze.stats.distributions.Rand.VariableSeed.randBasis

  "A Poisson(10.0)" should "have mean 10.0" in {
    import breeze.stats.distributions.Poisson
    val p = Poisson(10.0)
    val m = p.mean
    assert(math.abs(m - 10.0) < 0.000001)
  }




JAX/ml-ga.py
#!/usr/bin/env python3
# ml-ga.py
# Maximum likelihood via gradient ascent

import os
import pandas as pd
import numpy as np
import scipy as sp

import jax
from jax import grad, jit
import jax.numpy as jnp
import jax.scipy as jsp

print("Maximum likelihood by gradient ascent using JAX")

print("First read and process the data (using regular Python)")
df = pd.read_csv(os.path.join("..", "pima.data"), sep=" ", header=None)
print(df)
n, p = df.shape
print(n, p)

y = pd.get_dummies(df[7])["Yes"].to_numpy(dtype='float32')
X = df.drop(columns=7).to_numpy()
X = np.hstack((np.ones((n,1)), X))
print(X)
print(y)

print("Now gradient ascent using JAX")
X = X.astype(jnp.float32)
y = y.astype(jnp.float32)

@jit
def ll(beta):
    return jnp.sum(-jnp.log(1 + jnp.exp(-(2*y - 1)*jnp.dot(X, beta))))

gll = jit(grad(ll)) # use auto-diff for the gradient

@jit
def one_step(b0, learning_rate=1e-6):
    return b0 + learning_rate*gll(b0)

def ascend(step, init, max_its=10000, tol=1e-5, verb=True):
    def term(state):
        x1, x0, its = state
        return ((its > 0) & jnp.logical_not(jnp.allclose(x1, x0, tol)))
    def step_state(state):
        x1, x0, its = state
        x2 = step(x1)
        return [x2, x1, its - 1]
    b_opt, _, its_remaining = jax.lax.while_loop(
        term, step_state, [init, -init, max_its])
    if (verb):
        print(str(its_remaining) + " iterations remaining")
    return(b_opt)

init = jnp.array([-9.8, 0.1, 0, 0, 0, 0, 1.8, 0]).astype(jnp.float32)
print(init)
print(ll(init))
opt = ascend(one_step, init)
print(opt)
print(ll(opt))
print("Goodbye.")


# eof



Scala/src/main/scala/ML-GA.scala
/*
ML-GA.scala

Simple gradient ascent algorithm for maximum likelihood estimation
of a logistic regression model, applied to the Pima data

*/

import breeze.linalg.*
import breeze.numerics.*
import smile.data.pimpDataFrame
import annotation.tailrec

type DVD = DenseVector[Double]

object GradientAscent:


  @main def run() =
    println("First read and process the data")
    val df = smile.read.csv("../pima.data", delimiter=" ", header=false)
    print(df)
    val y = DenseVector(df.select("V8").
      map(_(0).asInstanceOf[String]).
      map(s => if (s == "Yes") 1.0 else 0.0).toArray)
    println(y)
    val x = DenseMatrix(df.drop("V8").toMatrix.toArray:_*)
    println(x)
    val ones = DenseVector.ones[Double](x.rows)
    val X = DenseMatrix.horzcat(ones.toDenseMatrix.t, x)
    println(X)
    val p = X.cols
    println(p)

    println("Now define log likelihood and gradient")
    def ll(beta: DVD): Double =
      sum(-log(ones + exp(-1.0*(2.0*y - ones)*:*(X * beta))))
    def gll(beta: DVD): DVD =
      (X.t)*(y - ones/:/(ones + exp(-X*beta)))

    println("Now define a functions for gradient ascent")
    def oneStep(learningRate: Double)(b0: DVD): DVD =
      b0 + learningRate*gll(b0)
    def ascend(step: DVD => DVD, init: DVD, maxIts: Int = 10000,
        tol: Double = 1e-8, verb: Boolean = true): DVD =
      @tailrec def go(b0: DVD, ll0: Double, itsLeft: Int): DVD =
        if (verb)
          println(s"$itsLeft : $ll0")
        val b1 = step(b0)
        val ll1 = ll(b1)
        if ((math.abs(ll0 - ll1) < tol)|(itsLeft < 1))
          b1
        else
          go(b1, ll1, itsLeft - 1)
      go(init, ll(init), maxIts)

    println("Now run a simple gradient ascent algorithm")
    // Better choose a reasonable init as gradient ascent is terrible...
    val init = DenseVector(-9.8, 0.1, 0, 0, 0, 0, 1.8, 0)
    val opt = ascend(oneStep(1e-6), init)
    println("Inits: " + init)
    println("Init ll: " + ll(init))
    println("Opt: " + opt)
    println("Opt ll: " + ll(opt))
    println("Goodbye.")



Haskell/app/Main.hs
{-# LANGUAGE ConstraintKinds, DataKinds, FlexibleContexts, GADTs,
 OverloadedStrings, PatternSynonyms, QuasiQuotes,
 ScopedTypeVariables, TemplateHaskell, TypeOperators, TypeApplications,
 ViewPatterns #-}

module Main (main) where

import Frames
import Frames.TH (rowGen, RowGen(..))
import Pipes hiding (Proxy)
import Numeric.LinearAlgebra
import qualified Data.Foldable as F
import Lens.Micro.Extras

-- template Haskell to create the Person type, and personParser
tableTypes' (rowGen "../pima.data")
            { rowTypeName = "Person"
            , columnNames = [ "npreg", "glu", "bp"
                            , "skin", "bmi", "ped", "age", "yy" ]
            , separator = " " }

-- create a data stream
dataStream :: MonadSafe m => Producer Person m ()
dataStream = readTableOpt personParser "../pima.data"

-- load full dataset
loadData :: IO (Frame Person)
loadData = inCoreAoS dataStream

-- create rows of covariate matrix
rec2l :: Person -> [Double]
rec2l r = [1.0, fromIntegral $ rgetField @Npreg r, fromIntegral $ rgetField @Glu r,
           fromIntegral $ rgetField @Bp r, fromIntegral $ rgetField @Skin r,
            rgetField @Bmi r, rgetField @Ped r, fromIntegral $ rgetField @Age r]

-- sum an hmatrix Vector
vsum :: Vector Double -> Double
vsum v = (konst 1 (size v) :: Vector Double) <.> v

-- log-likelihood
ll :: Matrix Double -> Vector Double -> Vector Double -> Double
ll x y b = (negate) (vsum (cmap log (
                              (scalar 1) + (cmap exp (cmap (negate) (
                                                         (((scalar 2) * y) - (scalar 1)) * (x #> b)
                                                         )
                                                     )))))

-- gradient
gll :: Matrix Double -> Vector Double -> Vector Double -> Vector Double
gll x y b = (tr x) #> (y - (scalar 1)/((scalar 1) + (cmap exp (-x #> b))))

-- one step of gradient ascent
oneStep :: Matrix Double -> Vector Double -> Double -> Vector Double -> Vector Double
oneStep x y lrate b0 = b0 + (scalar lrate) * (gll x y b0)

-- function for ascent
ascend :: (Vector Double -> Vector Double) ->
  Vector Double -> Int -> Double -> Vector Double
ascend astep init0 maxIts tol =
    go init0 maxIts
  where 
    go b0 itsLeft =
      let b1 = astep b0
          d = norm_2 $ b1 - b0
          in if ((d < tol)||(itsLeft < 1)) then b1
                else go b1 (itsLeft - 1)
    
main :: IO ()
main = do
  putStrLn "Gradient ascent for a log likelihood"
  putStrLn "First load and process the data..."
  dat <- loadData
  let yl = (\x -> if x then 1.0 else 0.0) <$> F.toList (view yy <$> dat)
  let xl = rec2l <$> F.toList dat
  print $ head xl
  let y = vector yl
  print y
  let x = fromLists xl 
  disp 2 x
  putStrLn "Now run the gradient ascent"
  -- choose reasonable init, since gradient ascent is terrible...
  let init0 = fromList [-9.8, 0.1, 0, 0, 0, 0, 1.8, 0] :: Vector Double
  print init0
  print $ ll x y init0
  let opt = ascend (oneStep x y 1.0e-6) init0 10000 1.0e-5
  print opt
  print $ ll x y opt
  putStrLn "Goodbye."


